### DISCUSSION TITLE
Auto save form data

### DISCUSSION THREAD

**Discussion by rikiless** [filamentphp/filament#5549] [Help]
Is it possible to create form with auto-saving/submitting values to Eloquent record?

I use Filament admin panel for building forms and using some reactive components for dependency fields too.

**Comment by SidneySaints** [comment]
as workaround maybe you can use `afterStateUpdated` method on form fields and call save method

**Reply by rikiless** [reply]
Ok. I found that following code is working but I need to re-apply this for every form field I have in form. It's not very smooth and also I need to cooperate with form validation.

```php

Forms\Components\TextInput::make('personal_number')
                    ->maxLength(255)
                    ->reactive()
                    ->afterStateUpdated(function (?Model $record, Field $component, $state) {
                        $record->{$component->getName()} = $state;
                        $record->save();
                    }),
```

**Reply by SidneySaints** [reply]
jep, it isn smart in any way. thats why we call it "workaround" 😄 . What we need is maybe something like
```php
$form->afterStateUpdated()
```

to cooperate with validation you can define your own custom auto save method that includes your own validation. You can get all rules that exists in form by:
```php 
$this->getRules()
``` 

in any way, its not smart but i think it could work

**Reply by SidneySaints** [reply]
Something like this:
```php
public function autoSave():void{
    $validator = Validator::make(['data'=>$this->form->getState()], $this->getRules())
        if(!count($validator->invalid())){
            $this->save();
        }
}
```

**Reply by rikiless** [reply]
Thanks. I will try it.

**Reply by enshtein** [reply]
> Thanks. I will try it.

did you get autosave?

**Comment by chasecmiller** [comment]
Hey guys, I'm late to the game here, but I'm writing what I did to make this work. 

On your EditRecord page, add a method called updated();
```
         /**
	 * @param $propertyName
	 */
	public function updated($propertyName) : void {
		$data = $this->validateOnly($propertyName);
		$data = $data['data'];
		dd('it worked, do what you need to do.',$propertyName, $data);
	}
```

On your form inputs, add the ->live() call.

Now, once you update the field, it will validate it and send it to the updated method and you can handle it as needed.  It's just using existing Livewire live update / validation.

**Reply by krekas** [reply]
Don't use `live()` like that. It will make way to many server requests and db updates. At least add a debounce

**Reply by chasecmiller** [reply]
Of course.  They can use live how they choose, I'm just showing how to do it without a workaround.  It's impossible for me to guess the time someone should set their debounce to, or if they should go with onBlur.  It's an outline of how to do it, that's all.

**Reply by alanost** [reply]
I am currently building a wizard and want to save the answers after each step.

I implemented the following code:

```php
public function updated($propertyName): void
{
    logger()->info('Property updated:', [
        'property' => $propertyName,
        'data' => $this->form->getState()
    ]);

    // save() without Redirect and show notification to count updated
    $this->save(shouldRedirect: false, shouldSendSavedNotification: true);
}
```

However, the issue is that when I change 4 fields and click on next, the `save` function is triggered 4 times (I see it because I get 4 notifications).

<img width="405" alt="Bildschirmfoto 2025-01-21 um 10 54 37" src="https://github.com/user-attachments/assets/a5bf3abc-5ead-4448-a46a-0a5314a9b8e8" />

Is there an alternative method that allows me to save only once when moving to the next step in the wizard when editing a record? Is there a different hook than `updated`? I’m new to Filament and Laravel and am unsure where to look to find the available API functions. Could someone kindly point me in the right direction? Thanks

**Comment by chasecmiller** [comment]
There are other ways.  Are you wanting to update the whole model on every character entry, or just a specific column in the db?  

If you are sure about doing the whole model instead of just the current field, use ->live(onBlur: true) or ->live(debounce: timeinseconds) to debounce it or defer it at least.  Then it won't save a ton of times.  Otherwise, I'd directly just update the model and save the complex queries and remove the notification all together.
